### PR TITLE
[ENG-1152] feat(devnet):  Harden devnet setup validation and source provenance

### PR DESCRIPTION
**Pull request by zelenevn**
## Summary

Harden the devnet setup flow (ENG-1152) so an invalid transition config fails
fast — before any expensive image build or the long-running Toccata/KIP-21
rehearsal.

**Preflight library.** New `scripts/lib/devnet-preflight.sh`, sourced by
`setup-devnet.sh` and run before `docker compose build`/`up`. Functions are pure
(inputs via args/env) so they unit-test without Docker:

- `resolve_devnet_env` — load a single effective env source (`.env` if present,
  otherwise the committed `.env.devnet.example`) plus `versions.devnet.env`, so
  build-time and runtime inputs match. Shell/CLI values win over the file for
  `FINALITY_PERIOD_SECONDS`, `TOCCATA_ACTIVATION_DAA_SCORE` and `IGRA_LANE_ID`
  (an explicitly empty value opts out).
- `validate_devnet_env` — aggregate every config problem into one actionable
  report: NETWORK, TX_ID_PREFIX, IGRA_LANE_ID (format + required when Toccata is
  scheduled), host ports incl. `KASWALLET_HOST_PORT` (+ collisions), mining
  address/threads, lock-script pair, finality range, DAA ordering and
  genesis-timestamp positivity.
- `mining_preflight` — when Toccata is scheduled, validate the mining config
  (address/threads) and remind the operator to point a miner at the kaspad gRPC
  port.
- `record_source_revisions` — write a timestamped branch/SHA/dirty manifest for
  the four built repos (rusty-kaspa-private, reth-private, kaswallet,
  igra-rpc-provider) to a gitignored `rehearsals/`, pruned to the newest
  `REHEARSAL_KEEP` (default 20) so each rehearsal is reproducible.

**Standalone miner.** kaspa-miner is not part of the compose stack;
`scripts/dev/run-devnet-miner.sh` clones, builds and runs `tmrlvi/kaspa-miner`
on the host against the devnet kaspad gRPC port. Mining is required to reach
`TOCCATA_ACTIVATION_DAA_SCORE` for the KIP-21 rehearsal.

**Compose hardening.** Operator-controlled values are passed to kaspad as
runtime env (referenced as `$VAR`) rather than spliced in at compose-render
time, with in-entrypoint `IGRA_LANE_ID` hex/length validation as a backstop for
a direct `docker compose up`. RPC is loopback + read-only by default
(`RPC_BIND_ADDR=127.0.0.1`, `RPC_READ_ONLY=true`) with an unsafe-exposure
warning; `READ_ONLY`, `RPC_PORT` and the kaswallet host port have safe compose
defaults. Build-time and runtime env come from one resolved source.

## Testing

- `./scripts/dev/test-devnet-preflight.sh` — 58/58 passing
- `shellcheck` clean on `setup-devnet.sh`, `devnet-preflight.sh`,
  `test-devnet-preflight.sh`, `run-devnet-miner.sh`
- Integration smoke: an invalid config exits non-zero with aggregated errors
  before any docker call.

### CODE REVIEW

File: scripts/setup-devnet.sh [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+# Fail fast on bad config BEFORE any expensive build/up.
+validate_devnet_env || exit 1
+
+# Record exactly what will be built (one manifest per rehearsal).
+record_source_revisions "$PROJECT_DIR/build/repos" "$PROJECT_DIR/rehearsals"
+
+# Mining is mandatory for the KIP-21 rehearsal whenever Toccata is scheduled.
+if [ -n "${TOCCATA_ACTIVATION_DAA_SCORE:-}" ]; then
+    mining_preflight "$PROJECT_DIR/build/repos" "$PROJECT_DIR/build/Dockerfile.kaspa-miner" || exit 1
```
Comment (pycckuu) [inline scripts/setup-devnet.sh]: **Issue**: With the default `TOCCATA_ACTIVATION_DAA_SCORE=1000`, `mining_preflight` runs and requires both `build/repos/kaspa-miner` and `build/Dockerfile.kaspa-miner`; `build_devnet_stack` (line 141) additionally lists `kaspa-miner` in its mandatory source-repo loop. But `scripts/dev/setup-repos.sh` (the documented clone step) only clones reth, kaswallet, igra-rpc-provider and kaspad — never kaspa-miner — and `build/Dockerfile.kaspa-miner` no longer exists (removed when the kaspa-miner service was dropped from the orchestra). kaspa-miner was intentionally removed from the repo and `setup-repos.sh` does not clone it, so a fresh `./scripts/setup-devnet.sh` aborts before bringing up the stack: with the default Toccata config it fails in `mining_preflight` ("kaspa-miner Dockerfile not found", "kaspa-miner source repo not found"), and even with Toccata disabled it fails in the build loop ("source repos not found: kaspa-miner"). The documented happy path cannot complete.

**Suggestion**: Since the miner is intentionally not part of the repo and operators run it separately, drop the kaspa-miner coupling from setup rather than re-adding it: remove `kaspa-miner` from the `build_devnet_stack` required-repos loop, and drop the repo/Dockerfile checks from `mining_preflight` (keep only the mining-address/threads validation, or remove the preflight entirely). Operators point an external miner at the published kaspad gRPC port (`127.0.0.1:${KASPAD_GRPC_PORT}`). Update the README clone/mining instructions accordingly, and optionally have setup print a reminder that reaching `TOCCATA_ACTIVATION_DAA_SCORE` requires the operator to start a miner.

File: versions.devnet.env [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Docker image versions for Devnet deployments.
+#
+# The four core services are built from source (see scripts/setup-devnet.sh). The
+# :devnet tags name the locally-built images so they never clobber pulled tags.
+KASPAD_VERSION=devnet
+RETH_VERSION=devnet
+KASWALLET_VERSION=devnet
+RPC_PROVIDER_VERSION=devnet
```
Comment (pycckuu) [inline versions.devnet.env:8]: **Issue**: `run_setup` (in `scripts/lib/setup-common.sh`) calls `validate_required_image_versions`, which `die`s unless `KASPAD_VERSION`, `RETH_VERSION`, `RPC_PROVIDER_VERSION`, `KASWALLET_VERSION`, `NODE_HEALTH_CHECK_VERSION` and `ATAN_UPLOADER_VERSION` are all non-empty. This file defines only the first four, and `.env.devnet.example` sets neither of the remaining two, so a devnet run aborts with "Required image versions are missing or still TODO placeholders: NODE_HEALTH_CHECK_VERSION ATAN_UPLOADER_VERSION" — after the slow build, before any container starts — even though the devnet compose runs neither service. Both `versions.mainnet.env` and `versions.galleon-testnet.env` define these two keys.

**Suggestion**: Add `NODE_HEALTH_CHECK_VERSION` and `ATAN_UPLOADER_VERSION` to `versions.devnet.env` (any non-empty, non-`TODO_` value passes, since the devnet stack doesn't run those services), or make `validate_required_image_versions` only require versions for the services actually present in the resolved compose file.

File: docker-compose.devnet.yml [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+          fi
+
+  kaspa-miner:
+    # "mining" profile so setup does not start it before it is built. Run manually:
+    #   docker compose --profile mining up -d --build kaspa-miner
+    profiles: ["mining"]
+    restart: *service-restart-policy
+    build:
+      context: build/repos/kaspa-miner
+      dockerfile: ../../Dockerfile.kaspa-miner
```
Comment (pycckuu) [inline docker-compose.devnet.yml]: **Issue**: The `kaspa-miner` service builds from `context: build/repos/kaspa-miner` with `dockerfile: ../../Dockerfile.kaspa-miner` (i.e. `build/Dockerfile.kaspa-miner`). Neither path exists — kaspa-miner was intentionally removed from the repo (`build/` contains only `.dockerignore`, and `setup-repos.sh` does not clone it). The README's documented manual step `docker compose --profile mining up -d --build kaspa-miner` therefore fails with a missing build context / Dockerfile, so the Toccata/KIP-21 rehearsal — which needs mined blocks to reach the activation DAA score — cannot proceed.

**Suggestion**: Since the miner is no longer part of the repo, remove the `kaspa-miner` service and the `mining` profile from this file (or replace the `build:` with a prebuilt external `image:`), and update the README "Mining" section to tell operators to run their own miner against the devnet kaspad gRPC endpoint (`127.0.0.1:${KASPAD_GRPC_PORT}`) instead of building one from the repo.

File: docker-compose.devnet.yml [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        - default
+    image: igranetwork/kaswallet:${KASWALLET_VERSION}
+    restart: *service-restart-policy
+    container_name: kaswallet-0-devnet
+    profiles: ["frontend-w1"]
+    logging: *default-logging
+    expose:
+      - "8082"
+    ports:
+      - "127.0.0.1:8082:8082"  # open 8082 on localhost for entry transactions
```
Comment (pycckuu) [inline docker-compose.devnet.yml]: **Issue**: `kaswallet-0` publishes a hardcoded host port `"127.0.0.1:8082:8082"`. Every other host-published port in this file (`KASPAD_*_PORT`, `EL_RPC_HOST_PORT`, `EL_WS_HOST_PORT`, `RPC_PORT`) is parameterized specifically so the devnet can coexist with a production/testnet stack on one host, and `validate_devnet_env` range- and collision-checks all of them. Port 8082 is neither parameterized nor in that collision set, yet the production kaswallet also binds 8082 — so the stated coexistence goal breaks on a silent host-port clash that the validation otherwise claims to prevent (and `RPC_PORT=8082` would pass validation but fail at `docker compose up`).

**Suggestion**: Parameterize the host side, e.g. `"${KASWALLET_HOST_PORT:-8082}:8082"`, add `KASWALLET_HOST_PORT` to `.env.devnet.example`, and include it in the port set validated/collision-checked by `validate_devnet_env`.

File: docker-compose.devnet.yml [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    environment:
+      RUST_LOG: ${RUST_LOG}
+      SERVER_HOST: "0.0.0.0"
+      SERVER_PORT: 8535
+      EL_URL: http://execution-layer:8545
+      EL_WS_URL: ws://execution-layer:8546
+      SECURITY_ENABLE_WHITELIST: "true"
+      # Off-box write exposure is prevented by RPC_READ_ONLY (default true) and
+      # the loopback host bind below; flip RPC_READ_ONLY to enable writes.
+      READ_ONLY: ${RPC_READ_ONLY}
```
Comment (pycckuu) [inline docker-compose.devnet.yml]: **Issue**: `READ_ONLY: ${RPC_READ_ONLY}` has no default, so if `RPC_READ_ONLY` is unset — e.g. a `docker compose -f docker-compose.devnet.yml up` that bypasses `setup-devnet.sh` with a `.env` missing the key — it renders as an empty string rather than the documented safe default of `true`. The adjacent comment claims "RPC_READ_ONLY (default true)", but that default lives only in the `.env.devnet.example` template, not in the compose substitution. The dependency-wait (`--tcp-if READ_ONLY false ...`) and the healthcheck also branch only on the literal string `false`, so an empty value skips the kaswallet readiness gate and leaves the wallet-backed write path's safety dependent on the provider's internal default.

**Suggestion**: Make the default explicit and fail-safe: `READ_ONLY: ${RPC_READ_ONLY:-true}`. This guarantees read-only when the variable is absent and keeps the rendered value consistent with the documented default and the dependency/healthcheck branches.

File: README.md [hunk 6]
```diff
@@ -63,6 +63,64 @@ For detailed guides, see:
...
+
+**Reset.** Consensus override params (`finality_depth`, `toccata_activation`) are
+baked into kaspad's consensus DB on first run, so changing `FINALITY_PERIOD_SECONDS`
+or `TOCCATA_ACTIVATION_DAA_SCORE` requires a fresh kaspad volume. To wipe all state:
+
+```bash
+# remove containers AND the kaspad_data named volume (destroys the chain):
+docker compose -f docker-compose.devnet.yml down -v
+# remove host-side state created by setup-devnet.sh (reth data is root-owned):
+sudo rm -rf data network-params logs overrides
```
Comment (pycckuu) [inline README.md]: **Issue**: `record_source_revisions` writes one timestamped manifest per run into `rehearsals/` (gitignored) and never prunes, but the documented "wipe all state" reset is `sudo rm -rf data network-params logs overrides` — it omits `rehearsals/`. So manifests accumulate unbounded across runs and survive a documented full reset, leaving stale state the operator isn't told about (and two runs within the same UTC second overwrite each other's manifest).

**Suggestion**: Add `rehearsals` to the reset command, or state that manifests are intentionally retained as an audit trail and cap retention in `record_source_revisions` (e.g. keep the newest N).

File: scripts/lib/devnet-preflight.sh [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+# Functions are pure (inputs via args/env, output via message + return code) so
+# they can be unit-tested without Docker (see scripts/dev/test-devnet-preflight.sh).
+
+# --- predicate helpers (return 0 = true, 1 = false; no output) ---
+
+is_uint()          { [[ "${1:-}" =~ ^[0-9]+$ ]]; }
+is_positive_int()  { [[ "${1:-}" =~ ^[0-9]+$ ]] && (( ${1:-0} > 0 )); }
+is_hex()           { [[ "${1:-}" =~ ^[0-9a-fA-F]+$ ]]; }
+is_hex_even()      { [[ "${1:-}" =~ ^([0-9a-fA-F]{2})+$ ]]; }
+is_port()          { [[ "${1:-}" =~ ^[0-9]+$ ]] && (( ${1:-0} >= 1 && ${1:-0} <= 65535 )); }
```
Comment (pycckuu) [inline scripts/lib/devnet-preflight.sh]: **Issue**: The integer predicates accept zero-padded values that bash arithmetic then misinterprets as octal. `is_uint "08"` matches the regex, and `is_port`/`is_positive_int` then evaluate `(( 08 ... ))`, which prints `bash: ((: 08: value too great for base (error token is "08")` to stderr; `"011"` is silently read as octal 9, giving a wrong comparison with no error. In `validate_devnet_env`, the line `local genesis_ts=$(( launch / 10 ... ))` (line 118) hits the same error for a zero-padded launch score, and because a `local` declaration masks the failing exit status the script does not abort under `set -e` — `genesis_ts` is left unset and the positivity check misbehaves silently.

**Suggestion**: Force base 10 in every arithmetic context, e.g. `(( 10#${1:-0} >= 1 && 10#${1:-0} <= 65535 ))` in the predicates and `10#$launch` / `10#$l1daa` in the genesis-timestamp computation, or tighten `is_uint` to reject leading zeros (`^(0|[1-9][0-9]*)$`). Add a test feeding a zero-padded value.

File: scripts/dev/test-devnet-preflight.sh [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+NETWORK=devnet
+FINALITY_PERIOD_SECONDS=600
+RPC_PORT=8555
+ENVEOF
+cat > "$RESOLVE_TMP/versions.devnet.env" <<'ENVEOF'
+KASPAD_VERSION=devnet
+ENVEOF
+
+ok "resolve: loads template when no .env" bash -c "$(declare -f resolve_devnet_env); resolve_devnet_env '$RESOLVE_TMP' .env.devnet.example versions.devnet.env; [ \"\$RPC_PORT\" = 8555 ] && [ \"\$KASPAD_VERSION\" = devnet ] && [ \"\$DEVNET_ENV_SOURCE\" = '$RESOLVE_TMP/.env.devnet.example' ]"
+ok "resolve: shell override wins"         bash -c "$(declare -f resolve_devnet_env); FINALITY_PERIOD_SECONDS=999 resolve_devnet_env '$RESOLVE_TMP' .env.devnet.example versions.devnet.env; [ \"\$FINALITY_PERIOD_SECONDS\" = 999 ]"
```
Comment (pycckuu) [inline scripts/dev/test-devnet-preflight.sh:119]: **Issue**: `resolve_devnet_env`'s subtlest behavior — preserving an explicitly empty shell override so it wins over a non-empty file value (the reason it uses `${!k+x}` rather than `${k:-}`) — is never tested. This is exactly the documented "set `TOCCATA_ACTIVATION_DAA_SCORE=` (empty) to opt out" / `IGRA_LANE_ID` opt-out contract. The only override test sets a non-empty value (`FINALITY_PERIOD_SECONDS=999`), which would also pass under a naive implementation, so it does not guard the set-but-empty path.

**Suggestion**: Add a case where the template sets `IGRA_LANE_ID=97b10000` (or `TOCCATA=1000`) and the call passes `IGRA_LANE_ID=` (empty), asserting the resolved value is empty; and a symmetric case proving an unset shell var falls through to the file value.

File: docker-compose.devnet.yml [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+      KASWALLET_HOST: kaswallet-0
+      WALLET_DAEMON_URI: http://kaswallet-0:8082
+      WALLET_TO_ADDRESS: ${W0_WALLET_TO_ADDRESS}
+      KASWALLET_PASSWORD: ${W0_KASWALLET_PASSWORD}
+    expose:
+      - "8535"
+    ports:
+      # Loopback by default (no Traefik/TLS/auth on devnet). Set RPC_BIND_ADDR=0.0.0.0
+      # only to deliberately expose the wallet-backed RPC to the LAN/internet.
+      - "${RPC_BIND_ADDR:-127.0.0.1}:${RPC_PORT}:8535"
```
Comment (pycckuu) [inline docker-compose.devnet.yml]: **Issue**: The rpc-provider host-port mapping uses `${RPC_PORT}` with no fallback, while `setup-devnet.sh`'s `print_summary` and the README use `${RPC_PORT:-8555}`. If `.env` omits `RPC_PORT`, `docker compose up` fails with an empty-port mapping error instead of using the 8555 default referenced everywhere else.

**Suggestion**: Use `${RPC_PORT:-8555}` in the mapping to match the documented default and keep the compose file self-consistent when `RPC_PORT` is unset.

File: scripts/lib/devnet-preflight.sh [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        fi
+    fi
+
+    # DAA ordering
+    for name in IGRA_LAUNCH_DAA_SCORE L1_REFERENCE_DAA_SCORE L1_REFERENCE_TIMESTAMP; do
+        is_uint "${!name:-}" || \
+            errors+=("$name must be a non-negative integer (got: '${!name:-<unset>}')")
+    done
+    if is_uint "$launch" && is_uint "$l1daa" && is_uint "$l1ts"; then
+        local genesis_ts=$(( launch / 10 - l1daa / 10 + l1ts - 1 ))
```
Comment (pycckuu) [inline scripts/lib/devnet-preflight.sh]: **Issue**: The EL-genesis-timestamp guard computes `launch / 10 - l1daa / 10 + l1ts - 1` with two independent integer divisions. Per-term truncation differs from truncating the difference: for DAA scores that aren't multiples of 10, the result can differ by 1 from `(launch - l1daa)/10 + l1ts - 1`. If `run-igra-el.sh` derives the genesis timestamp by subtracting before dividing, this guard can accept a config the EL rejects (or vice versa) right at the `genesis_ts > 0` boundary. (With the shipped defaults of 0/0 the two agree.)

**Suggestion**: Confirm the exact rounding `run-igra-el.sh` uses and mirror it here. If it subtracts first, use `local genesis_ts=$(( (launch - l1daa) / 10 + l1ts - 1 ))`.

File: docker-compose.devnet.yml [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+        if [ "$$NETWORK" != "mainnet" ]; then
+          ARGS="--$$NETWORK --nodnsseed $$ARGS"
+        fi
+
+        IGRA_ENABLED="$${IGRA_ENABLE:-true}"
+        if [ "$$IGRA_ENABLED" = "true" ]; then
+          ARGS="$$ARGS \
+            --igra-enable \
+            --atan-listen \
+            --atan-transaction-id-prefix=${TX_ID_PREFIX} \
```
Comment (pycckuu) [inline docker-compose.devnet.yml]: **Issue**: Operator-controlled `.env` values (`TX_ID_PREFIX`, `IGRA_LAUNCH_DAA_SCORE`, `IGRA_LOCK_SCRIPT_PUBKEY`, `GENESIS_BLOCK_HASH`, `L1_REFERENCE_*`, etc.) are substituted at compose-render time directly into the shell-evaluated kaspad entrypoint, and the unquoted `exec ... /app/kaspad $ARGS` means a value containing whitespace injects extra kaspad flags while a value containing `$(...)` executes in the container at arg-build time. `validate_devnet_env` constrains these only when the operator runs `setup-devnet.sh`; a direct `docker compose up` skips it. The production `docker-compose.yml` keeps in-entrypoint fail-fast checks and `${VAR:?}` guards for the same values; this devnet file omits them.

**Suggestion**: Mirror the production backstops here — add `${VAR:?}` guards to the kaspad `environment:` entries and/or pass these values as runtime `environment:` entries referenced as `$$VAR` (as `NETWORK`/`IGRA_LANE_ID` already are) rather than splicing compose-time `${VAR}` into the entrypoint, so an attacker- or typo-supplied value is data rather than shell text.

**Review by pycckuu — CHANGES_REQUESTED** [review]
A couple of blocking items to sort out before this comes up cleanly — see comments below.

Also, please address the conflicts.
